usaspending_api/references/v2/views/naics.py

```diff
@@ -0,0 +1,178 @@
+import logging
+from collections import OrderedDict
+
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from django.db.models.functions import Length
+
+from usaspending_api.common.cache_decorator import cache_response
+from usaspending_api.common.validator.tinyshield import TinyShield
+from usaspending_api.references.models import NAICS
+
+logger = logging.getLogger("console")
+
+
+class NAICSViewSet(APIView):
+    """
+    Return a list of NAICS or a filtered list of NAICS
+    """
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/v2/references/naics.md"
+
+    def _parse_and_validate_request(self, requested_naics: str, request_data) -> dict:
+        naics_filter = request_data.get("filter")
+        data = {"code": requested_naics, "description": naics_filter}
+        models = [
+            {"key": "code", "name": "code", "type": "integer", "allow_nulls": True, "optional": True},
+            {
+                "key": "description",
+                "name": "description",
+                "type": "text",
+                "text_type": "search",
+                "default": None,
+                "optional": True,
+                "allow_nulls": True,
+            },
+        ]
+        validated = TinyShield(models).block(data)
+        return validated
+
+    def _fetch_children(self, naics_code, naics_filter=None) -> list:
+        results = []
+        if naics_filter:
+            naics = NAICS.objects.annotate(text_len=Length("code")).filter(
+                **naics_filter, code__startswith=naics_code, text_len=len(naics_code) + 2
+            )
+        else:
+            naics = NAICS.objects.annotate(text_len=Length("code")).filter(
+                code__startswith=naics_code, text_len=len(naics_code) + 2
+            )
+        for naic in naics:
+            if len(naic.code) < 6:
+                result = OrderedDict()
+                result["naics"] = naic.code
+                result["naics_description"] = naic.description
+                result["count"] = (
+                    NAICS.objects.annotate(text_len=Length("code"))
+                    .filter(code__startswith=naic.code, text_len=6)
+                    .count()
+                )
+            else:
+                result = OrderedDict()
+                result["naics"] = naic.code
+                result["naics_description"] = naic.description
+                result["count"] = 0
+            results.append(result)
+        return results
+
+    def _filter_search(self, naics_filter: dict) -> dict:
+        tier1_codes = set([])
+        tier2_codes = set([])
+        tier3_codes = set([])
+
+        tier3_naics = NAICS.objects.annotate(text_len=Length("code")).filter(**naics_filter, text_len=6)
```

Comment:
Perhaps obtain all of the desired naics in a single DB query and group them by "level" in code instead of calling the DB several times to construct the objects?

Reply:
sure